### PR TITLE
expression,ddl: check table name for columns in partition expression

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

Fix https://github.com/pingcap/tidb/issues/8647

### What is changed and how it works?

Before the change in this line https://github.com/pingcap/tidb/compare/master...tiancaiamao:issue-8647?expand=1#diff-619bc716bb72d6b9ad7c92317f7c6434R109

`ParseSimpleExprWithTableInfo` will only compare the column field, it will ignore table name field, so this SQL could success (the bug):

```
create table xxx ... partition by range (yyy.purchased)
``` 

Other changes are irrelevant and they are here just to make CI pass

### Check List 

Tests 

 - Unit test

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @crazycs520 @XuHuaiyu

**Comment by tiancaiamao** [comment]
PTAL @crazycs520 @XuHuaiyu

File: ddl/db_partition_test.go [hunk 1]
```diff
@@ -233,6 +233,19 @@ func (s *testIntegrationSuite9) TestCreateTableWithPartition(c *C) {
...
 	// TODO: fix this one
 	// assertErrorCode(c, tk, `create table t33 (a timestamp, b int) partition by hash(unix_timestamp(a)) partitions 30;`, tmysql.ErrPartitionFuncNotAllowed)
+
+	// Fix issue 8647
+	assertErrorCode(c, tk, `CREATE TABLE trb8 (
+		id int(11) DEFAULT NULL,
+		name varchar(50) DEFAULT NULL,
+		purchased date DEFAULT NULL
+	) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin
+	PARTITION BY RANGE ( year(fuck.purchased) - 1 ) (
```
Comment (crazycs520) [inline ddl/db_partition_test.go]: `fuck`? Excuse me?

File: ddl/partition.go [hunk 2]
```diff
@@ -474,7 +474,7 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
 func checkRangePartitioningKeysConstraints(sctx sessionctx.Context, s *ast.CreateTableStmt, tblInfo *model.TableInfo, constraints []*ast.Constraint) error {
 	// Returns directly if there is no constraint in the partition table.
 	// TODO: Remove the test 's.Partition.Expr == nil' when we support 'PARTITION BY RANGE COLUMNS'
-	if len(constraints) == 0 || s.Partition.Expr == nil {
+	if s.Partition.Expr == nil {
```
Comment (crazycs520) [inline ddl/partition.go]: Why remove "len(constraints) == 0"? 
Check table name for columns will be checked in `checkPartitionFuncType`.

Comment (tiancaiamao) [reply ddl/partition.go]: I want line 484 to be executed, `checkPartitionFuncType` will return error but the error code is not correct
`extractPartitionColumns` in line 484 will return the correct error for this bug fix.

**Pull request merged**